src/bitcoin-chainstate.cpp

```diff
@@ -103,6 +103,11 @@ int main(int argc, char* argv[])
         {
             std::cout << "Warning: " << warning.original << std::endl;
         }
+        void fatalError(const std::string& debug_message, const bilingual_str& user_message) override
+        {
+            std::cerr << "Error: " << debug_message << std::endl;
+            std::cerr << (user_message.empty() ? "A fatal internal error occurred." : user_message.original) << std::endl;
```

Comment:
See the discussion in [this](https://github.com/bitcoin/bitcoin/pull/27636#pullrequestreview-1428804291) and [this](https://github.com/bitcoin/bitcoin/pull/27636#issuecomment-1550107951) comment. To summarize: Calling `std::abort` or `exit(0)` would skip over destructors and lead to an unclean exit. Functions calling `fatalError` should already be returning values that would lead to `bitcoin-chainstate` shutting down normally.

Reply:
Would suggest documenting this in `src/kernel/notifications_interface.h`. Maybe:

```c++
//! The fatal error notification is sent to notify the user and start
//! shutting down if an error happens in kernel code that can't be recovered
//! from. After this notification is sent, whatever function triggered the
//! error should also return an error code or raise an exception.
```